tutorials/graph-theory/prims-algorithm.md

```diff
@@ -9,4 +9,130 @@ keywords:
+
+### Complexity and Use-Cases
+When implemented using a priority queue (usually a binary heap), the time complexity of Prim’s algorithm is $O(E + V log V)$, where E is the number of edges and V is the number of vertices. This makes Prim's algorithm highly efficient for **dense** graphs, where the number of edges is close to the maximum possible. For sparse graphs (a graph with a relatively small number of edges compared to the maximum number of possible edges), other algorithms like [Kruskal’s](kruskals-algorithm.md) may perform better.
+
+The space complexity of the Prims Algorithm is $O(V + E)$, where V is the number of vertices and E is the number of edges.
+
+Prim's algorithm is widely used in network design, where the goal is to minimize the cost of building a network while connecting all nodes. Examples include laying cables, designing road networks, or creating efficient communication paths. It is also used as a building block in other algorithms for solving more complex problems.
+
+![Prims Algorithm Image](https://miro.medium.com/max/700/1*7kpPIPcmXr38Juh0umM6fA.jpeg)
+
+## Example [1135 - Connecting Cities With Minimum Cost](https://leetcode.com/problems/connecting-cities-with-minimum-cost/description/?envType=problem-list-v2&envId=minimum-spanning-tree)
+**Note:** The above link requires the Leetcode Subscription. To view the problem, use [this](https://leetcode.ca/all/1135.html) alternate link.
+
+### Instructions: 
+- There are N cities numbered from 1 to N.
+- You are given connections, where each connections[i] = [city1, city2, cost] represents the cost to connect city1 and city2 together. The connections are all bidirectional.
+- Return the minimum cost so that for every pair of cities, there exists a path of connections (possibly of length 1) that connects those two cities together.  The cost is the sum of the connection costs used. If the task is impossible, return -1.
+
+Example: 
+**Input:** N = 3, connections = [[1,2,5],[1,3,6],[2,3,1]]
+**Output:** 6
+
+### Approach
+First, we convert the given 'connections' array, into an adjacency list for ease of traversing.
+```python3
+connections = [[1,2,5],[1,3,6],[2,3,1]] # Example
+adjacency_list = {}
+
+for connection in connections:
+    if connection[0] not in adjacency_list:
+        adjacency_list[connection[0]] = []
+    if connection[1] not in adjacency_list:
+        adjacency_list[connection[1]] = []
+    adjacency_list[connection[0]].append((connection[2], connection[1]))
+    adjacency_list[connection[1]].append((connection[2], connection[0]))
+```
+The adjacency list will look like:
+```
+1: [(5, 2), (6, 3)]
+2: [(5, 1), (1, 3)]
+3: [(6, 1), (1, 2)]
+```
+Where each tuple consists of '(cost, destination)'.
+
+Now, to find the minimum cost, we use the below code.
+```python3
```

Comment:
1. let's put the step by step explanations first and show the full code at the end.
2. use the custom code block.  search `<Tabs>` in  kruskals page for reference.